### PR TITLE
ForLoopsLowering: Handle Iterable.withIndex() where the type is a bounded type parameter.

### PR DESCRIPTION
**Pull request by punzki**
TODO: Handle Sequences by extending DefaultIterableHandler.

### CODE REVIEW

**Comment by max-kammerer** [comment]
Thank you!

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt [hunk 1]
```diff
@@ -731,23 +731,26 @@ internal class WithIndexHandler(context: CommonBackendContext, private val visit
...
-                            it.valueParameters.isEmpty()
-                }).apply {
-                    dispatchReceiver = expression
-                }
+            // The lowering operates on subtypes of Iterable. Therefore, the IrType could be
+            // a type parameter bounded by Iterable. When that is the case, we cannot get
+            // the class from the type and instead uses the Iterable.iterator() function.
+            val iteratorFun = expression.type.getClass()?.functions?.single {
+                it.name == OperatorNameConventions.ITERATOR &&
+                        it.valueParameters.isEmpty()
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt]: Shouldn't we also check that there's no extension receiver, or is it not relevant in this case?

Comment (punzki) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt]: Member function should always "win out" over any extension function, and there's always an `iterator()` member function on `Iterable`s.

Comment (udalov) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt]: I meant that a subclass of `Iterable` can have its own member extension iterator, which could lead to `single` throwing an exception here. Something like this:
```
class MyIterable<T> : Iterable<T> {
    ....
    override fun iterator(): Iterator<T> = ...

    // Not an iterator, although has the name "iterator" and no value parameters!
    fun SomeUnrelatedType.iterator(): Double = ...
}
```

Comment (udalov) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt]: @punzki I've been investigating a related problem in our project bootstrap with JVM IR, and fixed it along with this one here, please take a look: https://upsource.jetbrains.com/kotlin/review/KOTLIN-CR-3586

**Pull request merged**